### PR TITLE
provider: make :ruby provider check use same code path as :python

### PR DESCRIPTION
**Pull request by mhinz**
Change the macro to return the cached has_<provider> variables right away. This
way we can easily put code behind it that is only run once for initialization.

Prior to this change, the provider check for ruby was special-cased and now it
returns E319, like :python etc.

References #9354

cc @justinmk

### CODE REVIEW

File: src/nvim/eval.c [hunk 1]
```diff
@@ -22752,6 +22741,8 @@ bool eval_has_provider(const char *name)
                       sizeof("provider#" #name "#Call") - 1, \
                       false); \
       has_##name = !!find_func((char_u *)"provider#" #name "#Call"); \
+    } else { \
+      return has_##name; \
```
Comment (justinmk) [inline src/nvim/eval.c]: why in `else` ? Then it would return `false` on the first call , but not subsequent.

Anyways, this makes the ruby codepath awkward, so better not make this change. Doing (partial) control-flow in a macro can be hard to follow.

Comment (justinmk) [reply src/nvim/eval.c]: the outer `if (has_ruby == -1) {` isn't needed

Comment (mhinz) [reply src/nvim/eval.c]: It is needed If we don't change the macro to return early.

Assume that `CHECK_PROVIDER(ruby)` sets it to 1. Then the additional code would be run every time as long as `provider#ruby#Detect()` returns 1 as well.

Comment (justinmk) [reply src/nvim/eval.c]: can avoid the nesting like this:

```diff
+    bool need_check_ruby = (has_ruby == -1);
+    CHECK_PROVIDER(ruby);
+    if (need_check_ruby && has_ruby == 1) {
+      char *rubyhost = call_func_retstr("provider#ruby#Detect", 0, NULL, true);
+      if (rubyhost) {
+        if (*rubyhost == NUL) {
+          // Invalid rubyhost executable. Gem is probably not installed.
+          has_ruby = 0;
+        }
+        xfree(rubyhost);
+      }
+    }
```

Comment (mhinz) [reply src/nvim/eval.c]: Hmm, ok, technically the same. In that case I'd consider the slightly more nested version to be more readable.

Unless you're in strong favor of your version, I'd like to go with mine. :]

Comment (justinmk) [reply src/nvim/eval.c]: the nested version puts CHECK_PROVIDER in a different place than the other cases, which adds unnecessary questions that the reader needs to answer when comparing the code. and the unnested version assigns a variable with a descriptive name hinting at what's going on. how is that less readable?

Comment (mhinz) [reply src/nvim/eval.c]: More variables means more state to keep in mind. I'm not arguing though. ;]
